File: code/extract_revenues_sec_fin_stat_data.py
import pandas as pd

df_names = ["sub", "tag", "num", "pre"]
for df in df_names:
  globals()[df] = pd.read_csv('data/' + df + '.csv')

df_merged = sub[(sub.countryba == "US") & (sub.fp.str[0] == 'Q')]

df_merged = df_merged[['adsh', 'cik', 'name', 'sic', 'fp', 'period', 'fye']]
df_merged = df_merged[
  df_merged.groupby(['cik'])['period'].transform(max) == df_merged.period]

df_merged = pd.merge(df_merged, pre, on = ['adsh'], how = 'left')

df_merged = df_merged[(df_merged.stmt == "IS") & 
  ((df_merged.tag == "Revenues") | 
  (df_merged.tag == "RevenueFromContractWithCustomerExcludingAssessedTax") |
  (df_merged.tag == "RevenueFromContractWithCustomerIncludingAssessedTax"))]

df_merged = pd.merge(df_merged, num, on = ['adsh', 'tag', 'version'], how = 'left')

df_merged = df_merged[(df_merged.ddate == df_merged.period) & 
  (df_merged.uom == "USD") &
  (df_merged.qtrs == 1) & 
  df_merged.value.notnull() &
  df_merged.coreg.isna()]

df_merged = df_merged[
  df_merged.groupby(['adsh', 'cik'])['value'].transform(max) ==
  df_merged.value]
  
df_merged = df_merged.rename({'value': 'total_revenue'}, axis = 1)

df_merged = df_merged[['adsh', 'cik', 'name', 'sic', 'fp', 'fye', 
         'ddate', 'total_revenue']]

df_merged = df_merged.drop_duplicates(['cik', 'name', 'sic', 'fp', 
        'fye', 'ddate', 'total_revenue'])

print(len(df_merged.adsh))

File: code/extract_revenues_sec_fin_stat_data.R
library(tidyverse)
library(ExPanDaR)

for (df_name in c("sub", "tag", "num", "pre")) {
  assign(df_name, read_csv(paste0("data/", df_name, ".csv")))
}

sub %>%
  filter(countryba == "US",
         str_sub(fp, 1, 1) == "Q") %>%
  select(adsh, cik, name, sic, fp, period, fye) %>%
  group_by(cik) %>%
  top_n(1, wt = period) %>%
  left_join(pre) %>%
  filter(stmt == "IS",
         (tag == "Revenues") |
           (tag == "RevenueFromContractWithCustomerExcludingAssessedTax") |
           (tag == "RevenueFromContractWithCustomerIncludingAssessedTax")) %>%
  distinct(adsh, tag, version, .keep_all = TRUE) %>%
  left_join(num) %>%
  filter(ddate == period,
         qtrs == 1L,
         is.na(coreg),
         uom == "USD",
         !is.na(value)) %>%
  mutate(total_revenue = value)  %>%
  select(adsh, cik, name, sic, fp, fye,
         ddate, total_revenue) %>%
  group_by(adsh, cik, name, sic, fp, fye, ddate) %>%
  top_n(1, wt = total_revenue) %>%
  ungroup() %>%
  distinct(cik, name, sic, fp, fye, ddate, .keep_all = TRUE) %>%
  arrange(cik) -> revenue
  
nrow(revenue)

length(unique(revenue$cik)) == nrow(revenue)
table(revenue$ddate)
  
prepare_descriptive_table(revenue %>% select(-ddate, -cik))$df
  
revenue %>% arrange(-total_revenue)
  
#  ggplot(revenue) +
#    geom_density(aes(x = log(1 + total_revenue)),
#                 color = NA, fill = "lightblue") +
#    theme_minimal()
  
#  ggplot(revenue) +
#    geom_histogram(aes(x = log(1 + total_revenue)),
#                   color = NA, fill = "lightblue") +
#    theme_minimal()
